### PR TITLE
Fix slowdown of Cheetah 0.7.1 compared to 0.6.3

### PR DESCRIPTION
**Pull request by jank324**
## Description

Introduces a couple of changes that improve the speed of Cheetah by about 2x (measured on the ARES RL example):
 - Removes not-needed memory allocation in `Screen`.
 - More efficient transfer map reduction logic in `Segment.track` that avoids the use of `Segment.is_skippable`, which is more expensive than expected.
 - Makes more use of `register_buffer` and `register_parameter` in `__init__`s of beams and elements, because these are significantly more efficient that property assignments on `nn.Modules`, which are actually very slow.

## Motivation and Context

- [ ] I have raised an issue to propose this change (required for new features and bug fixes)

Running the ARES RL code again with a prototype of Cheetah 0.7.1, I found that the samples per second were reduced by a factor of about 3.5x when compared to 0.6.1 (which this code ran with originally), but also 0.6.3 (the last non-vectorised version.

This matches observations from the vectorisation PR #116, where at one point I wrote:

 > Looking at the optimize_speed.ipynb example Notebook, it seems like this change currently slows down Cheetah between 2x and 4x for single samples. On the other hand, if you run with 1,000 samples, we are looking at a 40x to 55x speed improvement on an M1 Pro CPU. This would likely be more with more samples and/or on GPU.

The goal of Cheetah is also to be fast, especially for the purpose of RL, so we should check if this can be fixed.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (update in the documentation)

## Checklist

- [ ] I have updated the changelog accordingly (**required**).
- [ ] My change requires a change to the documentation.
- [ ] I have updated the tests accordingly (_required for a bug fix or a new feature_).
- [ ] I have updated the documentation accordingly.
  
  
- [ ] I have reformatted the code and checked that formatting passes (**required**).
- [ ] I have have fixed all issues found by `flake8` (**required**).
- [ ] I have ensured that all `pytest` tests pass (**required**).
- [ ] I have run `pytest` on a machine with a CUDA GPU and made sure all tests pass (**required**).
- [ ] I have checked that the documentation builds (**required**).

Note: We are using a maximum length of 88 characters per line.

### CODE REVIEW

**Comment by jank324** [comment]
Some speed benchmarks:

### Env benchmark

```python
%%timeit

observation, info = env.reset()
done = False
while not done:
    observation, reward, terminated, truncated, info = env.step(env.action_space.sample())
    done = terminated or truncated
```

 - **0.6.1:** 65.4 ms ± 225 μs per loop (mean ± std. dev. of 7 runs, 10 loops each)
 - **0.7.1:** 166 ms ± 4.38 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

### Raw Cheetah benchmark

```python
segment.AREABSCR1.is_active = True
```

```python
%%timeit
outgoing = segment.track(incoming)
```

 - **0.6.1:** 722 μs ± 1.47 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
 - **0.7.1:** 1.01 ms ± 3.77 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)

**Comment by jank324** [comment]
It appears there are two main causes: `Segment.track` and `Screen.track`. In `Screen.track`, the slow part is a `torch.full` inside of `Screen.set_read_beam`, which creates a dummy screen image every time a new beam is read and the old screen image becomes invalid.

Screenshot 2025-03-07 at 09 47 04

I just ran some tests how this compares, for example, against setting `Screen.cached_reading` to `None`. Clearly this is the cause.

Screenshot 2025-03-07 at 09 49 05

The last commit therefore sets `Screen.cached_reading` to `None`. The resulting speeds are:

 - **Env benchmark:** 106 ms ± 18.4 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
 - **Raw Cheetah benchmark:** 1.11 ms ± 7.32 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)

**Comment by jank324** [comment]
It seems that `compute_relativistic_factors` and `base_rmatrix` are now the slowest parts.

Screenshot 2025-03-07 at 10 09 23

**Comment by jank324** [comment]
I just made a modification to the `__init__` of both beam classes that from the _Before_  profiling results I thought might improve the speed. According the the `timeit` benchmarks it didn't make any difference at all, but the profiling would suggest that it make quite a major difference ... I'm confused 🤔 

It would also have to be checked with @Hespe why the original order of buffer registration was introduced. Also, if we choose to keep the modification, it should be done in all `Element` subclasses as well.

 - **Env benchmark:** 96.9 ms ± 3.21 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
 - **Raw Cheetah benchmark:** 1.12 ms ± 2.46 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)

**Before**
Screenshot 2025-03-07 at 19 23 26

**After**
Screenshot 2025-03-07 at 19 23 54

**Comment by jank324** [comment]
Another realisation: The main expense in `Segment.track` seems to be caused by `base_rmatrix`, wherein the main expense is broadcasting. This then is just a price we have to pay for vectorising Cheetah I guess.

Screenshot 2025-03-07 at 19 35 17

For reference, here are a few experiments I did on the cost of running broadcasting in different ways by itself.

Screenshot 2025-03-07 at 19 36 34

File: cheetah/particles/parameter_beam.py [hunk 1]
```diff
@@ -37,18 +37,14 @@ def __init__(
...
-        self.energy = torch.as_tensor(energy, **factory_kwargs)
-        if total_charge is not None:
-            self.total_charge = torch.as_tensor(total_charge, **factory_kwargs)
+        total_charge = total_charge if total_charge is not None else torch.tensor(0.0)
 
-        self.species = species if species is not None else Species("electron")
+        self.register_buffer("_mu", mu.to(**factory_kwargs))
+        self.register_buffer("_cov", cov.to(**factory_kwargs))
+        self.register_buffer("energy", energy.to(**factory_kwargs))
+        self.register_buffer("total_charge", total_charge.to(**factory_kwargs))
```
Comment (Hespe) [inline cheetah/particles/parameter_beam.py:47]: The trouble with this style of initialization is that `nn.Parameter` are not properly registered since all properties are forced as buffers. That is, if you pass a `nn.Parameter` while creating a beam, torch will silently convert the parameter to a buffer. Creating a parametrized beam is still possible, but only _after_ initialization by assigning a `nn.Parameter` to a property.

Comment (Hespe) [reply cheetah/particles/parameter_beam.py:47]: So essentially, the previous way of initialization was a way to take advantage of property assignment magic that torch is performing in the background. But to be honest, I am not sure if the initial manual registration did anything useful or if we could just rely on the property assignment as long as we make sure to always assign tensors.

Comment (jank324) [reply cheetah/particles/parameter_beam.py:47]: If we just assign a tensor as a property, it actually won't be registered as a buffer. I will run some speed tests with and without the change and then probably revert it. The property assignment magic seems to take some time ... but it is a feature that we want.

And either way, I'm still confused about whether the change actually made a difference or not. The measurements are contradictory 🤔

**Comment by Hespe** [comment]
Would be interesting to see how much we pay for broadcasting if all tensors are scalars.

**Comment by jank324** [comment]
I just ran the benchmarks again

 - **Env benchmark:** 91.8 ms ± 630 μs per loop (mean ± std. dev. of 7 runs, 1 loop each)
 - **Raw Cheetah benchmark:** 950 μs ± 10.9 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)

**Comment by jank324** [comment]
And here another run with the parameter registration change undone:

 - **Env benchmark:** 109 ms ± 833 μs per loop (mean ± std. dev. of 7 runs, 1 loop each)
 - **Raw Cheetah benchmark:** 1.28 ms ± 22 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)

**Comment by Hespe** [comment]
Ok, so it's a roughly 20% increase that we have to pay in these two cases if we want to be able to assign `nn.Parameter`s at initialization. Not really that great. I assume modying the beam to pass it along instead of allocating a new object is not an option?

**Comment by jank324** [comment]
> Ok, so it's a roughly 20% increase that we have to pay in these two cases if we want to be able to assign `nn.Parameter`s at initialization. Not really that great. I assume modying the beam to pass it along instead of allocating a new object is not an option?

Hmm ... so from a user experience point of view I never liked that idea. It invites programming errors if your incoming beam is different after tracking. On the other hand, reserving new memory does introduce a non-neglible overhead.

I'm thinking right now if introducing an `inplace` argument to `track`, which is `False` by default, but elements internally might set it to `True` and this way they reuse the beam object, but the user never sees this (unless the user wants to but then its their own fault).

**Comment by jank324** [comment]
I also just added the profiling outputs for both cases as well.

**Comment by Hespe** [comment]
> I'm thinking right now if introducing an `inplace` argument to `track`, which is `False` by default, but elements internally might set it to `True` and this way they reuse the beam object, but the user never sees this (unless the user wants to but then its their own fault).

With the idea being `Segment` does so, except for the very first element? Such that the user does not have the incoming beam modified, but internally we do not have to allocate new tensors for each element? I guess that would not even be noticable from the outside if all elements properly implement that behaviour.

**Comment by jank324** [comment]
One more note. In the profiler output it seems that the more recent ones are about 10x - 20x slower ... this is simply because I was using 10th of the sampling interval at one point.

**Comment by jank324** [comment]
> > I'm thinking right now if introducing an `inplace` argument to `track`, which is `False` by default, but elements internally might set it to `True` and this way they reuse the beam object, but the user never sees this (unless the user wants to but then its their own fault).
> 
> With the idea being `Segment` does so, except for the very first element? Such that the user does not have the incoming beam modified, but internally we do not have to allocate new tensors for each element? I guess that would not even be noticable from the outside if all elements properly implement that behaviour.

It could probably be even be implemented in the `Element` base class. I guess the general template would be:

```python
def track(self, incoming: Beam, inplace: bool = True) -> Beam:
    outgoing = incoming if inplace else incoming.clone()
    
    outgoing.particles = # Do computations ...
```

**Comment by Hespe** [comment]
> It could probably be even be implemented in the `Element` base class. I guess the general template would be:
> 
> ```python
> def track(self, incoming: Beam, inplace: bool = True) -> Beam:
>     outgoing = incoming if inplace else incoming.clone()
>     
>     outgoing.particles = # Do computations ...
> ```
> 
> On the other hand, `ParameterBeam.__setattr__` appears to be the truly expensive part ... and that would remain.

Oh yes, you are right. In place tensor operations on the other hand would be troublesome here, right? They would be incompatible with automatically broadcasting within the elements. And could lead to problems with differentiation?

**Comment by jank324** [comment]
I didn't necessarily mean inplace tensor operations actually. What I mean was more along these lines ... the first is what `Element` currently does, the second would be what it could do in an inlplace case.

Screenshot 2025-03-10 at 12 01 37

Btw ... I discovered a bug, where the species is not passed in `Element.track`.

**Comment by jank324** [comment]
According to the benchmark I just posted, the inplace operation would be faster. We should consider this for the future, but for the specific ARES RL example that I need to speed up right now, the code actually combines all transfer maps, and then does only one multiplication. Even in the suggested case there would have to be at least one new beam copy created, and here it only creates that one. So this optimisation would not give an advantage in this case. Similarly the copy for `Screen.read_beam` slows things down, but that one is unavoidable as well.

The only other source of slowdown as far as I can tell is the broadcasting in `base_rmatrix` ... we need that too. Regarding this, I'm not sure @Hespe what you meant with

 > Would be interesting to see how much we pay for broadcasting if all tensors are scalars.

The example I'm running uses a vectorised version of Cheetah (i.e. >=0.7.0), but non of the inputs have a vector dimension.

**Comment by Hespe** [comment]
> The only other source of slowdown as far as I can tell is the broadcasting in `base_rmatrix` ... we need that too. Regarding this, I'm not sure @Hespe what you meant with
> 
> > Would be interesting to see how much we pay for broadcasting if all tensors are scalars.
> 
> The example I'm running uses a vectorised version of Cheetah (i.e. >=0.7.0), but non of the inputs have a vector dimension.

I was referring to your micro benchmark above, that just does some broadcasting of tensors `a` and `b` of different shapes. If they have different shapes, the cost of broadcasting is somehow unavoidable (not sure actually, operations that internally broadcast without explicit call to `broadcast_tensors` are likely very efficient). If `broadcast_tensors` still takes 2 us for scalar tensors, that could add quite a bit of time to some scalar simulations.

**Comment by jank324** [comment]
Here is the same benchmark for scalar tensors. Other broadcasting operations must do the same thing, 

Screenshot 2025-03-10 at 13 07 34

For the fun of it, I just ran a broadcasting math operation. This is clearly faster. I'm not sure why, but my best guess is that the broadcasting here is done in the compiled code rather than the interpreted Python.

Screenshot 2025-03-10 at 13 10 46

Screenshot 2025-03-10 at 13 11 09

Maybe this means we should look through the uses of manual broadcasting operations and see if they can be replaced with other operations. My assumption is that they cannot. Specifically the one in `base_rmatrix` would have a significant impact on the speed of Cheetah.

**Comment by jank324** [comment]
@Hespe and I just went through everything and optimised every last bit on the ARES example that we think can be optimised.

The current numbers are:
 - **Env benchmark:** 95.8 ms ± 1.62 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
 - **Raw Cheetah benchmark:** 1.06 ms ± 11.2 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)

**Comment by jank324** [comment]
The question came up if `is_skippable` should become a function to denote that computing it is expensive ... a lot more expensive than we thought.

**Comment by jank324** [comment]
I just reran everything with the most recent commit, to make sure that the changes didn't negatively affect anything.

The current numbers are:
 - **Env benchmark:** 85.8 ms ± 1.59 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
 - **Raw Cheetah benchmark:** 926 μs ± 13.3 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)

These were all run on an Apple M1 Pro. We also ran a test earlier on a Dell Windows laptop, and saw that compute times were about 1.7x longer.

**Comment by jank324** [comment]
For reference: The RL training went from 1,450 fps with v0.6.1 to 1,343 fps with v0.7.1 ... so less than 8% worse, which I'm pretty happy with.

File: cheetah/accelerator/bpm.py [hunk 2]
```diff
@@ -28,8 +28,11 @@ def __init__(
         super().__init__(name=name, device=device, dtype=dtype)
 
         self.is_active = is_active
+
         self.register_buffer(
-            "reading", torch.tensor(torch.nan, device=device, dtype=dtype)
+            "reading",
+            torch.as_tensor(torch.nan, device=device, dtype=dtype),
```
[the remark this answers is not in the document]
Comment (jank324) [reply cheetah/accelerator/bpm.py:34]: I suppose

File: cheetah/accelerator/dipole.py [hunk 3]
```diff
@@ -78,41 +78,58 @@ def __init__(
...
+                **factory_kwargs,
+            ),
+        )
+        self.register_buffer_or_parameter(
+            "fringe_integral_exit",
+            torch.as_tensor(
+                (
+                    fringe_integral_exit
+                    if fringe_integral_exit is not None
+                    else self.fringe_integral
```
[the remark this answers is not in the document]
Comment (jank324) [reply cheetah/accelerator/dipole.py:115]: I put all of them fully inside this now ... if it's already a tensor with the right `dtype`, `as_tensor` won't do much at all.

File: cheetah/accelerator/dipole.py [hunk 4]
```diff
@@ -78,41 +78,58 @@ def __init__(
...
+            ),
+        )
+
+        self.register_buffer_or_parameter(
+            "gap", torch.as_tensor(gap if gap is not None else 0.0, **factory_kwargs)
+        )
+        self.register_buffer_or_parameter(
+            "gap_exit",
+            torch.as_tensor(
+                gap_exit if gap_exit is not None else self.gap, **factory_kwargs
```
[the remark this answers is not in the document]
Comment (jank324) [reply cheetah/accelerator/dipole.py:127]: Same answer

File: cheetah/accelerator/screen.py [hunk 5]
```diff
@@ -67,28 +67,42 @@ def __init__(
...
+                **factory_kwargs,
+            ),
+        )
+        self.register_buffer_or_parameter(
+            "kde_bandwidth",
+            torch.as_tensor(
+                (
+                    kde_bandwidth
+                    if kde_bandwidth is not None
+                    else self.pixel_size[0].clone()
```
[the remark this answers is not in the document]
Comment (jank324) [reply cheetah/accelerator/screen.py:89]: Huh?

[the remark this answers is not in the document]
Comment (Hespe) [reply cheetah/accelerator/screen.py:89]: Same thing as above, the `torch.as_tensor` could go inside.

File: cheetah/accelerator/bpm.py (hunk 2 above)
Comment (jank324) [inline cheetah/accelerator/bpm.py:34]:
```suggestion
            torch.as_tensor((torch.nan, torch.nan), device=device, dtype=dtype),
```

[the remark this answers is not in the document]
Comment (jank324) [reply cheetah/accelerator/bpm.py:34]: Added as suggestion

File: tests/test_screen.py [hunk 6]
```diff
@@ -149,13 +148,92 @@ def test_reading_dtype_conversion():
...
+    the incoming out outgoing beam after tracking. Considers the case where the incoming
+    beam is a `ParameterBeam`.
+    """
+    incoming = cheetah.ParameterBeam.from_astra(
+        "tests/resources/ACHIP_EA1_2021.1351.001"
+    )
+    screen = cheetah.Screen(is_active=True)
+
+    outgoing = screen.track(incoming)
+    original_read_beam = screen.get_read_beam()
```
Comment (jank324) [inline tests/test_screen.py:172]:
```suggestion
    original_read_beam = screen.get_read_beam().clone()
```

File: tests/test_screen.py [hunk 7]
```diff
@@ -149,13 +148,92 @@ def test_reading_dtype_conversion():
...
+    the incoming out outgoing beam after tracking. Considers the case where the incoming
+    beam is a `ParticleBeam`.
+    """
+    incoming = cheetah.ParticleBeam.from_astra(
+        "tests/resources/ACHIP_EA1_2021.1351.001"
+    )
+    screen = cheetah.Screen(is_active=True)
+
+    outgoing = screen.track(incoming)
+    original_read_beam = screen.get_read_beam()
```
Comment (jank324) [inline tests/test_screen.py:209]:
```suggestion
    original_read_beam = screen.get_read_beam().clone()
```

**Review by cr-xu — APPROVED** [review]

**Comment by cr-xu** [comment]
@jank324  @Hespe  On a side note, it would be helpful to gather the insights & experience here into a written document (e.g. to `clone` or not, when to broadcast, what operations can be substituted by more efficient ones...) so that in the future new contributors can also implement the elements in a consistent & fast-executing way.
